baseband/mark4/frame.py

```diff
@@ -141,35 +142,91 @@ def fromdata(cls, data, header=None, verify=True, **kwargs):
-        data = np.empty(self.shape, self.dtype)
-        if self.valid:
-            valid_start = self.shape[0] * VALIDSTART // PAYLOADSIZE
-            data[:valid_start] = self.invalid_data_value
-            data[valid_start:] = self.payload.data
-        else:
-            data[...] = self.invalid_data_value
-        return data
-
-    @property
-    def shape(self):
-        """Shape of the data held in the payload (samples_per_frame, nchan)."""
-        return (self.payload.shape[0] * PAYLOADSIZE //
-                (PAYLOADSIZE - VALIDSTART),) + self.payload.shape[1:]
+    def samples_per_frame(self):
+        """Number of samples (including those overwritten by header)."""
+        return self._valid_start + self.payload.nsample
 
     def __getitem__(self, item=()):
         if isinstance(item, six.string_types):
             return self.header.__getitem__(item)
+
+        # Normally, we would just pass on to the payload here, but for
+        # Mark 4, we need to deal with data overwritten by the header.
+        valid_start = self._valid_start
+        if item is () or item == slice(None):
+            # Short-cut for full payload.
+            if self.valid:
+                # Note: Creating an empty array and setting part to invalid
+                # is much faster than creating one pre-filled with invalid.
+                data = np.empty(self.shape, self.dtype)
+                data[:valid_start] = self.invalid_data_value
+                data[valid_start:] = self.payload.data
+                return data
+            else:
+                return np.full(self.shape, self.invalid_data_value, self.dtype)
+
+        if isinstance(item, tuple):
+            sample_index = item[1:]
+            item = item[0]
         else:
-            # Need to learn how to deal with invalid data part!  Hence,
-            # we cannot just slice the payload like vlbi_base.frame.
-            raise IndexError("{0} object can not be indexed or sliced yet."
-                             .format(type(self)))
+            sample_index = None
+
+        # Interpret item as an index or slice.
+        nsample = self.samples_per_frame
+        if not isinstance(item, slice):
+            try:
+                item = item.__index__()
+            except Exception:
+                raise TypeError("{0} object can only be indexed or sliced."
+                                .format(type(self)))
+            if item < 0:
+                item += nsample
+
+            if not (0 <= item < nsample):
+                raise IndexError("{0} index out of range.".format(type(self)))
+
+            payload_item = item - valid_start
+            if payload_item >= 0:
+                data = self.payload[payload_item]
+            else:
+                data = np.full(self.sample_shape, self.invalid_data_value,
+                               self.dtype)
+
+            return data if sample_index is None else data[sample_index]
+
+        # We have a slice.
```

Comment:
I prefer to write `# If instead we have a slice.` to make it explict to unobservant readers.